src/internal/clusterstate/v2.8.0/collection.go

```diff
@@ -0,0 +1,241 @@
+		params[indexFieldName(idx)] = idx.Extract(val)
+	}
+
+	return params, nil
+}
+
+// ErrNotFound indicates that a key was not found when it was expected to
+// exist.
+type ErrNotFound struct {
+	Type          string
+	Key           string
+	customMessage string
+}
+
+func (err ErrNotFound) Is(other error) bool {
+	_, ok := other.(ErrNotFound)
+	return ok
+}
+
+func (err ErrNotFound) Error() string {
+	if err.customMessage != "" {
+		return err.customMessage
+	}
+	return fmt.Sprintf("%s %s not found", strings.TrimPrefix(err.Type, DefaultPrefix), err.Key)
+}
+
+func (e ErrNotFound) GRPCStatus() *status.Status {
+	return status.New(codes.NotFound, e.Error())
+}
+
+// IsErrNotFound determines if an error is an ErrNotFound error
+func IsErrNotFound(err error) bool {
+	return errors.Is(err, ErrNotFound{})
+}
+
+// ErrExists indicates that a key was found to exist when it was expected not
+// to.
+type ErrExists struct {
+	Type          string
+	Key           string
+	customMessage string
+}
+
+func (err ErrExists) Is(other error) bool {
+	_, ok := other.(ErrExists)
+	return ok
+}
+
+const DefaultPrefix string = "pachyderm/1.7.0"
+
+func (err ErrExists) Error() string {
+	if err.customMessage != "" {
+		return err.customMessage
+	}
+	return fmt.Sprintf("%s %s already exists", strings.TrimPrefix(err.Type, DefaultPrefix), err.Key)
+}
+
+func (e ErrExists) GRPCStatus() *status.Status {
+	return status.New(codes.AlreadyExists, e.Error())
+}
+
+// IsErrExists determines if an error is an ErrExists error
+func IsErrExists(err error) bool {
+	return errors.Is(err, ErrExists{})
+}
+
+func (c *postgresCollection) withKey(key interface{}, query func(string) error) error {
+	if str, ok := key.(string); ok {
+		return query(str)
+	}
+	rawKey, err := c.keyGen(key)
+	if err != nil {
+		return err
+	}
+	err = query(rawKey)
+	if err != nil {
+		var notFound ErrNotFound
+		var exists ErrExists
+		if c.notFound != nil && errors.As(err, &notFound) {
+			notFound.customMessage = c.notFound(key)
+			return errors.EnsureStack(notFound)
+		}
+		if c.exists != nil && errors.As(err, &exists) {
+			exists.customMessage = c.exists(key)
+			return errors.EnsureStack(exists)
+		}
+		return err
+	}
+	return nil
+}
+
+func isDuplicateKeyError(err error) bool {
+	pqerr := &pq.Error{}
+	if errors.As(err, pqerr) {
+		return pqerr.Code == pgerrcode.UniqueViolation
+	}
+	return false
+}
+
+func (c *postgresCollection) mapSQLError(err error, key string) error {
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return errors.WithStack(ErrNotFound{Type: c.table, Key: key})
+		} else if isDuplicateKeyError(err) {
+			return errors.WithStack(ErrExists{Type: c.table, Key: key})
+		}
+		return errors.EnsureStack(err)
+	}
+	return nil
+}
+
+func migratePostgreSQLCollection(ctx context.Context, tx *pachsql.Tx, name string, indices []*index, oldVal proto.Message, f func(oldKey string) (newKey string, newVal proto.Message, err error), opts ...colOption) error {
```

Comment:
I know in the past, I've advocated for complete self isolation of migration code and avoid dependencies at all cost, but I've accepted the fact that it might not be feasible. I think in this case, it's safe to just import the stuff you implemented in 2.5.0 migration.